tree-sitter-stack-graphs/src/lib.rs

```diff
@@ -0,0 +1,620 @@
+//! and reference nodes, since the location information determines which parts of the source file
+//! the user can _click on_, and the _destination_ of any code navigation queries the user makes.
+//! To do this, add a `source_node` attribute, whose value is a syntax node capture:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "pop_symbol", symbol = (source-text @id), source_node = @func, is_definition
+//! }
+//! ```
+//!
+//! Note how in this example, we use a different syntax node for the _target_ of the definition
+//! (the entirety of the function definition) and for the _name_ of the definition (the content of
+//! the function's `name`).
+//!
+//! ### Connecting stack graph nodes with edges
+//!
+//! To connect two stack graph nodes, use the `edge` statement to add an edge between them:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "pop_symbol", symbol = (source-text @id), source_node = @func, is_definition
+//!   node body
+//!   edge def -> body
+//! }
+//! ```
+//!
+//! To implement shadowing (which determines which definitions are selected when multiple are available),
+//! you can add a `precedence` attribute to each edge to indicate which paths are prioritized:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "pop_symbol", symbol = (source-text @id), source_node = @func, is_definition
+//!   node body
+//!   edge def -> body
+//!   attr (def -> body) precedence = 1
+//! }
+//! ```
+//!
+//! (If you don't specify a `precedence`, the default is 0.)
+//!
+//! ### Referring to the singleton nodes
+//!
+//! The _root node_ and _jump to scope node_ are singleton nodes that always exist for all stack
+//! graphs.  You can refer to them using the `ROOT_NODE` and `JUMP_TO_SCOPE_NODE` global variables:
+//!
+//! ``` skip
+//! (function_definition name: (identifier) @id) @func {
+//!   node def
+//!   attr (def) type = "pop_symbol", symbol = (source-text @id), source_node = @func, is_definition
+//!   edge (ROOT_NODE -> def)
+//! }
+//! ```
+//!
+//! ## Using this crate from Rust
+//!
+//! If you need very fine-grained control over how to use the resulting stack graphs, you can
+//! construct and operate on [`StackGraph`][stack_graphs::graph::StackGraph] instances directly
+//! from Rust code.  You will need Rust bindings for the tree-sitter grammar for your source
+//! language — for instance, [`tree-sitter-python`][].  Grammar Rust bindings provide two global
+//! symbols that you will need: [`language`][] and [`STACK_GRAPH_RULES`][].
+//!
+//! [`tree-sitter-python`]: https://docs.rs/tree-sitter-python/*/
+//! [`language`]: https://docs.rs/tree-sitter-python/*/tree_sitter_python/fn.language.html
+//! [`STACK_GRAPH_RULES`]: https://docs.rs/tree-sitter-python/*/tree_sitter_python/constant.STACK_GRAPH_RULES.html
```

Comment:
Is it okay that this reference doesn't exist yet?

Reply:
Hm, good question! This is written with the idea that stack graph rules are automatically discovered as part of the language, and included in the crate. I don't know what the status of that is (@dcreager?), but I'll take this out because I expect we'll want to release before that is implemented.